### PR TITLE
Scope format! temporaries

### PR DESCRIPTION
**Pull request by jonhoo**
This places the temporaries that `format!` generates to refer to its arguments (through `&dyn Trait`) in a short-lived scope surrounding just the invocation of `format!`. This enables `format!` to be used in generators without the temporaries preventing the generator from being `Send` (due to `dyn Trait` not being `Sync`).

See rust-lang/rust#64477 for details.

### CODE REVIEW

**Comment by jonhoo** [comment]
See in particular @nikomatsakis' description of the problem in https://github.com/rust-lang/rust/issues/64477#issuecomment-534669068 and @Nemo157's proposed solution in https://github.com/rust-lang/rust/issues/64477#issuecomment-534712712 (which this PR implements).

File: src/test/ui/fmt/issue-64477.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// In khe past, the code generated by `format!` produced temporaries in the surrounding scope that
```
Comment (Centril) [inline src/test/ui/fmt/issue-64477.rs]:
```suggestion
// In the past, the code generated by `format!` produced temporaries in the surrounding scope that
```

**Comment by Centril** [comment]
This is a breaking change (that may only be noticed in the dynamic semantics rather than the static semantics). By inserting the `let` you are changing the drop order:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3231f995544ca8cf48045a108a28ac78

**Comment by jonhoo** [comment]
That is true — I hadn't considered that. I wonder if there's an easy way for us to just cause the created `&dyn Trait` items to be dropped, and not the inputs. I can't immediately think of one :thinking:

**Comment by jonhoo** [comment]
@Centril thinking some more about this, I don't think there's a way to get out of changing the drop order except by making `format!` actually parse its arguments and ensure that it takes a reference to each one somehow. I toyed around with something like
```rust
macro_rules! format {
    ($($arg:tt)*) => ((|| {$crate::fmt::format($crate::__export::format_args!($($arg)*))})())
}
```
But the closure tries to move arguments passed to `format!` without `&` (despite not having a `move` annotation), so that won't work.

In https://github.com/rust-lang/rust/issues/64477#issuecomment-534669068, @nikomatsakis said:
> Unfortunately, the `format!` temporaries include `dyn Trait` objects that are not `Send`. So the error is correct. The shorter-term fix would I think be to alter the desugaring of `format!` so that its intermediate values are `Send`, I'm afraid.

I'm not sure how we could go about doing that without adding additional bounds on types passed to `format!` (which would _definitely_ be a breaking change). If you have suggestions though, I'm all ears.

Ultimately, my guess at the moment is that we will either have to change drop order, have `format!` fully parse its arguments so it can drop only the references, or change the bounds of `format!`. I think the last one is out. The second one would probably be a fairly large change (possibly even making `format!` a proc-macro), and would be error-prone as we'd have to replicate the parsing from `format_args!` and then appropriately transform all inputs. The original suggestion seems like a neat compromise, but we would then have to decide whether the change in drop order is acceptable.

**Comment by Centril** [comment]
@craterbot run mode=check-only

**Comment by Centril** [comment]
@craterbot abort

**Comment by Centril** [comment]
@craterbot run mode=build-and-test

(Changes dynamic semantics possibly.)

**Comment by jonhoo** [comment]
@Centril the azure crater agent appears to be stuck, and the aws agent isn't picking up any new jobs?

**Comment by Centril** [comment]
cc @pietroalbini

**Comment by pietroalbini** [comment]
The Azure agent being stuck is expected, as we had problems with that VM over the weekend without actually time to investigate them. Not sure what happened with the AWS agent though, after a restart it started working again.

**Comment by nikomatsakis** [comment]
@jonhoo can you say a bit more about the specifics here? The temporaries that are being created -- what are their types? There may be room to make the analysis more precise. In particular, we know that `&dyn Foo` objects *don't have destructors* (and NLL already takes advantage of this), so maybe we don't have to worry about when the temporaries containing them are freed.

**Comment by nikomatsakis** [comment]
Actually, @jonhoo, don't answer that -- can you open a new issue describing the ergonomic problem and some of the details, and we can file it to be discussed *there*?

**Comment by jonhoo** [comment]
@nikomatsakis opened https://github.com/rust-lang/rust/issues/64960

**Comment by nikomatsakis** [comment]
@jonhoo thanks!

**Comment by Centril** [comment]
@jonhoo can you triage the regressions?

**Comment by jonhoo** [comment]
I'll do some digging and report back. A cursory reading suggests that `varkor.metasub.223fb77cb9866a58013d7eb7afab2d2711224d12` *might* be genuine (see [here](https://crater-reports.s3.amazonaws.com/pr-64856/try%234465d0049bf00b9224f1908770f277c145361aed/gh/varkor.metasub/log.txt)), but the others are unlikely to be.

**Comment by Centril** [comment]
@varkor Seems your proof assistant doesn't want to play ball. :)

**Comment by jonhoo** [comment]
Triage (WIP):
```
AttilaOrs._tempBug.7d087eed8cecac718445cba03999bbeedc5c1585
```
same spurious failure on current nightly

```
dobrakmato.rust-brainfuck.b9e6db9065dee7bb0ad5587763477c1f2264aad6
```
**cannot reproduce failure on nightly.**
manual inspection says format! isn't to blame.

```
duriantang.nickel_cors.5685cd9045bc264276689410a4052e3a8ed0a8e1
```
same spurious failure on current nightly

```
mantal.expert_system.873622fa400d5bd721e592f22561e98860f69536
```
**cannot reproduce failure on nightly.**
manual inspection says format! isn't to blame.

```
northpole79.tcp_comms.a943d47b88f905bf6ebe20c3d69d0308969d5739
```
same failure on current nightly

```
proudlygeek.emuchip-8.2f6e44576998b85f3dc3454fa50b51f03d33b1f1
```
same failure on current nightly

```
tommyjk.kv-server.61e87f3703ad4db2c0210a2ab04c0c52ea2a5745
```
same failure on current nightly

```
varkor.metasub.223fb77cb9866a58013d7eb7afab2d2711224d12
```
:warning: **genuine compile-time error**

```
autoperf-0.9.4
```
same failure on current nightly

```
bosun_emitter-1.2.0
```

```
carboxyl_time-0.0.3
```

```
extsort-0.2.0
```
bad extern crate

```
globwalk-0.7.1
```
bad extern crate

```
psyche-serde-0.2.19
```

```
rsst-0.1.0
```
unrelated (home dir)

```
s_app_dir-0.0.0
```
unrelated (home dir)

```
sodium-rust-1.0.1
```

```
strip-ansi-escapes-0.1.0
```
bad extern crate

**Comment by jonhoo** [comment]
@Centril based on ^ I'd say that @vorner's is the only regression.

**Comment by vorner** [comment]
:confused: I'm having trouble figuring what my regression you're referring to. Did you want to highlight me, or someone else?

**Comment by jonhoo** [comment]
@vorner sorry, I meant @varkor !

**Comment by jonhoo** [comment]
@Centril what are the next steps here?

**Comment by Centril** [comment]
^--- @rust-lang/libs

**Comment by dtolnay** [comment]
Regression minimized:

```rust
fn args(_: impl IntoIterator<Item = impl AsRef<str>>) {
    unimplemented!()
}

fn main() {
    let cow = std::borrow::Cow::Borrowed("");
    args(&[&cow, "", &String::new()]); // okay
    args(&[&cow, "", &{ String::new() }]); // not okay
}
```

```console
error[E0308]: mismatched types
 --> src/main.rs:8:25
  |
8 |     args(&[&cow, "", &{ String::new() }]); // not okay
  |                         ^^^^^^^^^^^^^ expected enum `std::borrow::Cow`, found struct `std::string::String`
  |
  = note: expected type `std::borrow::Cow<'_, str>`
             found type `std::string::String`
```

**Comment by JohnCSimon** [comment]
Pinging from triage
 Is this PR ready for the review process?
CC: @Centril @dtolnay @sfackler @varkor @jonhoo

Thanks

**Comment by jonhoo** [comment]
From my perspective, the PR is done and ready for review. The crater run revealed only one regression, and it is essentially the one @dtolnay gave above. Review will need to determine whether that's acceptable or not. I don't think there's a way to fix that one regression while also fixing the original problem the PR was designed to fix.

**Comment by KodrAus** [comment]
Personally, this change looks ok to me because that papercut of having to explicitly type the array or cast the first element in order to type check does already come up sometimes:

```rust
args(&[&cow as &str, "", &{ String::new() }]); // okay
```

```rust
let a: &[&str] = &[&cow as &str, "", &{ String::new() }]; // okay
args(a);
```

It seems like we’re working with odd behaviour on either side of the fence; unexpected long borrows on one side and inference issues with blocks on the other (does anybody know why `&{ String::new() }` is different from `&String::new()` here?

**Comment by KodrAus** [comment]
@rfcbot fcp merge

**Comment by JohnCSimon** [comment]
Ping from triage:
@jonhoo Seems we have a merge conflict. 
cc: @KodrAus @sfackler 
Thanks!

**Comment by sfackler** [comment]
@rfcbot fcp merge

**Comment by jonhoo** [comment]
@JohnCSimon conflict has now been resolved

**Comment by JohnCSimon** [comment]
Ping from triage:
@sfackler Merge conflict has been resolved, is this ready for review + merge?
cc: @KodrAus @jonhoo @dtolnay @Centril 
Thanks!

**Comment by sfackler** [comment]
Ping @Kimundi @SimonSapin @withoutboats

**Comment by SimonSapin** [comment]
Should `println!` and other macros that use `format_args!` do the same?

**Comment by sfackler** [comment]
`println` expressions aren't usually kept alive across a yield point, so I don't think it's an issue there like it is here.

**Comment by jonhoo** [comment]
@sfackler The tricky part is that all arguments to `println` are still borrowed within the same scope. For example, this also won't compile:
```rust
async fn foo() {}

fn bar() -> impl Send {
    async move {
        let x = std::rc::Rc::new(1);
        println!("{}", x);
        foo().await;
    }
}
```
We could do the same fix for `println` and friends, though I think that would warrant another crater run.

**Comment by sfackler** [comment]
That has nothing to do with `println` - if you remove the line it still doesn't compile. `x` itself lives past the await so the Rc has to be stored in the future.

**Comment by jonhoo** [comment]
@sfackler Ah, yes, of course, you're right. I don't think the value has to be included in the future though. Isn't the whole reason why this issue arises with `format!` in the first place that `format!` creates temporaries in the _surrounding_ scope that aren't `Send`? `println!` does the same, no?

**Comment by sfackler** [comment]
Not including `x` in the future would change its drop order with respect to normal synchronous functions, and there was an explicit decision to not differ there.

No, the problem with `format!` is that it introduces temporaries in the containing *expression*. If that expression contains a yield then you have a problem.

**Comment by JohnCSimon** [comment]
Ping from triage, this has sat idle for the last week  @jonhoo @sfackler

**Comment by brunoczim** [comment]
Do this mean we can use format! inside multi-threaded futures?

**Comment by sfackler** [comment]
You can already use format in multi-threaded features. You just can't put it in an expression with an `.await` without this change.

**Comment by brunoczim** [comment]
It is just that... this does not seem to work.

```rust
/*
[dependencies]

tokio = { version = "0.2.6", features = ["io-util", "io-std", "macros"] }
*/

use std::{fmt, future::Future, pin::Pin};
use tokio::io::{self, AsyncWriteExt};

pub trait AsyncWriteFmt {
    fn write_fmt<'this, 'args, 'async_trait>(
        &'this mut self,
        args: fmt::Arguments<'args>,
    ) -> Pin<Box<dyn Future<Output = io::Result<()>> + Send + 'async_trait>>
    where
        'this: 'async_trait,
        Self: Sync + 'async_trait;
}

impl AsyncWriteFmt for io::Stdout {
    fn write_fmt<'this, 'args, 'async_trait>(
        &'this mut self,
        args: fmt::Arguments<'args>,
    ) -> Pin<Box<dyn Future<Output = io::Result<()>> + Send + 'async_trait>>
    where
        'this: 'async_trait,
        Self: Sync + 'async_trait,
    {
        let string = format!("{}", args);
        drop(args);
        Box::pin(async move { self.write_all(string.as_bytes()).await })
    }
}

async fn write_hello() {
    write!(io::stdout(), "Hello, World!\n").await.unwrap()
}

fn main() {
    tokio::spawn(write_hello());
}
```

`write_hello` is not `Send + Sync` here. Because of this, it will fail with the following error on the last nightly:

```rust
error: future cannot be sent between threads safely
   --> src/main.rs:48:5
    |
48  |     tokio::spawn(write_hello());
    |     ^^^^^^^^^^^^ future returned by `write_hello` is not `Send`
    | 
   ::: /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.6/src/task/spawn.rs:123:21
    |
123 |         T: Future + Send + 'static,
    |                     ---- required by this bound in `tokio::task::spawn::spawn`
    |
    = help: within `core::fmt::Void`, the trait `std::marker::Sync` is not implemented for `*mut (dyn std::ops::Fn() + 'static)`
note: future is not `Send` as this value is used across an await
   --> src/main.rs:44:5
    |
44  |     write!(io::stdout(), "Hello, World!\n").await.unwrap()
    |     ---------------------------------------^^^^^^
    |     |
    |     await occurs here, with `$ crate :: format_args ! ($ ($ arg) *)` maybe used later
    |     has type `[std::fmt::ArgumentV1<'_>; 0]`
45  | }
    | - `$ crate :: format_args ! ($ ($ arg) *)` is later dropped here
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

**Comment by sfackler** [comment]
That error is coming from `write!`, not `format!`. `fmt::Arguments` are not Send, so any future containing them will not be Send.

**Comment by brunoczim** [comment]
Well, this seems to work, weirdly.
```rust
/*
[dependencies]

tokio = { version = "0.2.6", features = ["io-util", "io-std", "macros"] }
async-trait = "0.1.21"
*/

use async_trait::async_trait;
use std::{fmt, future::Future, pin::Pin};
use tokio::io::{self, AsyncWriteExt};

pub trait AsyncWriteFmt {
    fn write_fmt<'this, 'args, 'async_trait>(
        &'this mut self,
        args: fmt::Arguments<'args>,
    ) -> Pin<Box<dyn Future<Output = io::Result<()>> + Send + 'async_trait>>
    where
        'this: 'async_trait,
        Self: Sync + 'async_trait;
}

impl AsyncWriteFmt for io::Stdout {
    fn write_fmt<'this, 'args, 'async_trait>(
        &'this mut self,
        args: fmt::Arguments<'args>,
    ) -> Pin<Box<dyn Future<Output = io::Result<()>> + Send + 'async_trait>>
    where
        'this: 'async_trait,
        Self: Sync + 'async_trait,
    {
        let string = format!("{}", args);
        drop(args);
        Box::pin(async move { self.write_all(string.as_bytes()).await })
    }
}

async fn write_hello() {
    let mut stdout = io::stdout();
    let future = write!(stdout, "Hello, World!\n");
    future.await.unwrap()
}

fn main() {
    tokio::spawn(write_hello());
}
```

**Pull request merged**